src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetCoreMvc2Integration.cs

```diff
@@ -119,101 +77,96 @@ public AspNetCoreMvc2Integration(object actionDescriptorObj, object httpContextO
             {
-                if (_afterAction == null)
+                try
                 {
-                    Type type = actionDescriptor.GetType().Assembly.GetType("Microsoft.AspNetCore.Mvc.Internal.MvcCoreDiagnosticSourceExtensions");
-
-                    _afterAction = DynamicMethodBuilder<Action<object, object, object, object>>.CreateMethodCallDelegate(
-                        type,
-                        "AfterAction");
+                    target.Invoke(diagnosticSource, actionDescriptor, httpContext, routeData);
+                }
+                catch (Exception ex)
+                {
+                    // don't add this exception to span, it was not thrown by controller action,
+                    // profiled app should continue working
+                    Log.ErrorException("Error executing target method.", ex);
                 }
-            }
-            catch
-            {
-                // TODO: log this as an instrumentation error, we cannot call instrumented method,
-                // profiled app will continue working without DiagnosticSource
-            }
 
-            try
-            {
-                // call the original method, catching and rethrowing any unhandled exceptions
-                _afterAction?.Invoke(diagnosticSource, actionDescriptor, httpContext, routeData);
-            }
-            catch (Exception ex)
-            {
-                integration?.SetException(ex);
-                throw;
-            }
-            finally
-            {
-                integration?.Dispose();
+                // add tags that are only available after MVC action is executed
+                scope?.Span?.SetTag("http.status_code", httpContext?.Response.StatusCode.ToString());
             }
         }
 
-        /// <summary>
-        /// Tags the current span as an error. Called when an unhandled exception is thrown in the instrumented method.
-        /// </summary>
-        /// <param name="ex">The exception that was thrown and not handled in the instrumented method.</param>
-        public void SetException(Exception ex)
+        private static Scope CreateScope(object actionDescriptor, object httpContext)
         {
-            _scope?.Span?.SetException(ex);
+            string controllerName = actionDescriptor.GetProperty<string>("ControllerName").GetValueOrDefault();
+            string actionName = actionDescriptor.GetProperty<string>("ActionName").GetValueOrDefault();
+            string routeTemplate = actionDescriptor.GetProperty("AttributeRouteInfo").GetProperty<string>("Template").GetValueOrDefault();
+
+            var request = httpContext.GetProperty("Request");
+            string httpMethod = request.GetProperty<string>("Method").ToString().ToUpperInvariant();
+            string url = GetDisplayUrl(request).ToLowerInvariant();
+
+            Scope scope = Tracer.Instance.StartActive(OperationName);
```

Comment:
Same as other PR, there's potential here for a leak IF this is used in a using pattern, doesn't seem to be here, but the creation call above needs to be moved inside a using and adjust this, or a try/finally and handle